### ISSUE TITLE
Runtime hook for server startup/init event

### ISSUE THREAD

**Issue by septatrix** [unjs/nitro#1987]
### Describe the feature

It would be great if there were a possibility to run some code once during server startup. My potential use case would be running migrations programmatically. Currently there does not exist a hook for this.

It is unclear if running such code inside the `defineNitroPlugin` satisfies these requirements (run only once, run before any request is handled) and if yes, that should likely be documented.

PS: However, even if they do meet these requirements it does not seem like they support async calls.

```js
// is this guaranteed to run once, and before any request?
// PS: is async allowed here?
export default defineNitroPlugin(async () => await dbsession.migrate())
```

See also https://github.com/nuxt/nuxt/issues/22486

### Additional information

- [ ] Would you be willing to help implement this feature?

**Comment by botic** [comment]
A special hook running once at start up would be great to e.g. log some debugging information into the log stream of a Nuxt container. Imagine something like this:

```typescript
export default defineNuxtConfig({
  hooks: {
    start: (runtimeConfig) => {
      console.log(`Using Backend API Version: ${runtimeConfig.apiVersion}`);    
      console.log(`Feature Flag A: ${runtimeConfig.public.featureA}`);    
    }
  }
})
```

This would make it much easier to write out hints e.g. for a helpdesk team. They could just start with the log and look for the important configuration options.

**Comment by pi0** [comment]
@botic nitro plugins actually run once during server startup. what is current issue with Varaint a?

**Comment by pi0** [comment]
@septatrix

> It is unclear if running such code inside the defineNitroPlugin satisfies these requirements (run only once, run before any request is handled) and if yes, that should likely be documented.

 Yes, this is the behavior. Nitro plugins run once during server startup. It is [documented](https://nitro.unjs.io/guide/plugins) but PR welcome to improve and reword

 > PS: However, even if they do meet these requirements it does not seem like they support async calls.

You can track support for async plugins in https://github.com/unjs/nitro/issues/915 however generally I wouldn't recommend blocking (entire) server for tasks such as DB migrations. Instead you could create a utility like `ensureDBInitialized` in `utils/` and use it in the routes that need it or use a middleware.

[Nitro Tasks API](https://github.com/unjs/nitro/issues/1974) is also a relevant topic you might like to track.

**Comment by septatrix** [comment]
> Nitro plugins run once during server startup. It is [documented](https://nitro.unjs.io/guide/plugins) but PR welcome to improve and reword

I must have missed the "on the first" in the docs (- why does it even say "first", are there cases in which multiple initializations are possible?).

> > PS: However, even if they do meet these requirements it does not seem like they support async calls.
> 
> You can track support for async plugins in #915...

Thanks for the pointer, async plugins would be perfect for me than.

> ...however generally I wouldn't recommend blocking (entire) server for tasks such as DB migrations. Instead you could create a utility like `ensureDBInitialized` in `utils/` and use it in the routes that need it or use a middleware.
> 
> [Nitro Tasks API](https://github.com/unjs/nitro/issues/1974) is also a relevant topic you might like to track.

While performing migrations on startup might not sound ideal, it can often be the most practical solution and is widely used due to this. While it might not work for distributed stuff with all its complications, for single node services it is perfect. There is always a small downtime when restarting to the newer version, a few more seconds won't hurt. It is a lot easier to apply, instead of e.g. having to run commands inside a docker container.
And the overhead of having an `ensureMigrated` middleware is also not always wanted. I would have to somehow know which DB schema is currently deployed which either requires a DB read or file read, both of which would also be any and would then happen for every route which needs the DB (unless I start introducing global state within Nitro which one should also try to prevent). So for me running migrations on startup is the most practical solution.

**Comment by pi0** [comment]
Async plugins, if/when supported will likely show a warning if their evaluation takes too long.

Also for Nitro's built-in database API, lazy init will be used because you need the database once you need it(at least IMO). This kind of lazy startup is one of the core principles in Nitro's server design that ensures it is fast and scalable.

However, I respect your opinion about the convenience of blocking the whole server startup also it is nothing strange indeed many (traditional) approaches of DB migration block server startup.

I made an alternative discussion to support `init:` and `startup:` tasks we can manage them better at least with the same possibility of async plugins. So let's track it there.

**Comment by dervondenbergen** [comment]
Hello @pi0, colleague of @botic here.

Thanks for the information that Nitro Plugins only run once, we somehow missed this in the [docs which you linked](https://nitro.unjs.io/guide/plugins).

We use Nitro in the context of Nuxt, and tried our usecase there using all kind of hooks, both defined in the `nuxt.config.ts` and in a `defineNuxtPlugin`/`defineNitroPlugin`.

Not defining a hook and logging directly in the `defineNitroPlugin` callback as a server plugin finally worked 🎉 

<img width="332" alt="Bildschirmfoto 2023-12-23 um 16 41 14" src="https://github.com/unjs/nitro/assets/2054024/977d2ad1-886c-41a0-99b6-341a9d17a911">

**Comment by botic** [comment]
> @botic nitro plugins actually run once during server startup. what is current issue with Varaint a? (feel free to share feedback in a new issue if have issues or questions about it not working)

Yes, Variant A is perfect. I cannot reproduce my issue not having access to `useRuntimeConfig()` but since @dervondenbergen got it working, I'm totally fine 👌  

I personally was looking for an `init` or `start` hook at the beginning. My first thought was: If there is a `close` hook, where is its counterpart? But variant A is just fine and maybe just needs a hint / example in the documentation.

**Comment by septatrix** [comment]
> Also for Nitro's built-in database API, lazy init will be used because you need the database once you need it(at least IMO). This kind of lazy startup is one of the core principles in Nitro's server design that ensures it is fast and scalable.

How would such a lazy init work? The current nitro docs do not mention anything about database migrations and the [usage example](https://nitro.unjs.io/guide/database#usage) always calls `DROP TABLE IF EXISTS users` which is a no go. My initial guess was that it could be handled by the `useDatabase()` composable but that is synchronous so also likely not where such a lazy init would be applied. My final guess would be to always perform migrations before each ``await db.sql`SELECT ...`;`` but that also is dump because that would add at least a round trip to each database call because migration status would need to be checked.

My assumption is that this is not implemented, yet, however I am curious how you intend to integrate it.

**Comment by canidam** [comment]
I have a similar use-case but a bit different.
I have a long running task that needs to run on server startup. It's a background process that runs for the whole duration the app is running.
It fetch data from db (async), do some computations and updates. Currently, I was able to make it work with

`export default defineNitroPlugin(async () => { .. }`

I also noted it's not blocking other plugins (which is good for my usecase). Is that the way to implement that? Is there another hook / method I should consider?
